File: Assets/src/model/grasses/Bladegrass.cs
using System;
using System.Collections.Generic;
using System.Linq;
using UnityEngine;

[Serializable]
[ObjectInfo(description: "Walk over to sharpen.\nOnce sharpened, the first creature walking into this Bladegrass takes 2 damage and removes the Bladegrass.\nSharpened Bladegrass dies on its own after 10 turns.")]
public class Bladegrass : Grass, IActorEnterHandler, IActorLeaveHandler {
  public static Item HomeItem => new ItemBladegrassBlade();
  public static bool CanOccupy(Tile tile) => tile is Ground;
  public bool isSharp = false;
  [field:NonSerialized] /// controller only
  public event Action OnSharpened;
  public Bladegrass(Vector2Int pos) : base(pos) {}

  public void Sharpen() {
    if (!isSharp) {
      isSharp = true;
      OnSharpened?.Invoke();
      AddTimedEvent(10, KillSelf);
    }
  }

  public void HandleActorLeave(Actor obj) {
    Sharpen();
  }

  public void HandleActorEnter(Actor actor) {
    if (isSharp) {
      Kill(actor);
      actor.TakeDamage(2, this);
    }
  }
}

[Serializable]
[ObjectInfo("bladegrass")]
internal class ItemBladegrassBlade : EquippableItem, IWeapon {
  public (int, int) AttackSpread => (2, 2);
  public override EquipmentSlot slot => EquipmentSlot.Weapon;
  public override int stacksMax => int.MaxValue;
}

File: web/src/model/grasses/Bladegrass.ts
import { Grass } from './Grass';
import { ACTOR_ENTER_HANDLER, ACTOR_LEAVE_HANDLER, type IActorEnterHandler, type IActorLeaveHandler } from '../../core/types';
import { Ground } from '../Tile';
import { entityRegistry } from '../../generator/entityRegistry';
import type { Vector2Int } from '../../core/Vector2Int';
import type { Tile } from '../Tile';

/**
 * Sharpens when an actor leaves. When sharp, deals 2 damage on enter then dies.
 * Auto-dies after 10 turns once sharpened.
 * Port of C# Bladegrass.cs.
 */
export class Bladegrass extends Grass implements IActorEnterHandler, IActorLeaveHandler {
  readonly [ACTOR_ENTER_HANDLER] = true as const;
  readonly [ACTOR_LEAVE_HANDLER] = true as const;

  isSharp = false;
  onSharpened: (() => void) | null = null;

  constructor(pos: Vector2Int) {
    super(pos);
  }

  static canOccupy(tile: Tile): boolean {
    return tile instanceof Ground;
  }

  sharpen(): void {
    if (!this.isSharp) {
      this.isSharp = true;
      this.onSharpened?.();
      // this.addTimedEvent(10, () => this.killSelf());
    }
  }

  handleActorLeave(_actor: any): void {
    this.sharpen();
  }

  handleActorEnter(actor: any): void {
    if (this.isSharp) {
      this.kill(actor);
      actor.takeDamage(1, this);
    }
  }
}

entityRegistry.register('Bladegrass', Bladegrass);
